polkadot-parachains/parachains-common/src/impls.rs

```diff
@@ -113,6 +114,59 @@ impl<T: Get<MultiLocation>> FilterAssetLocation for AssetsFrom<T> {
 	}
 }
 
+/// Converter struct with two main usage scenarios:
+/// 1. Transfer asset from local to other para-chain
+/// 	- with `reverse_ref` to convert a numeric asset ID (must be `TryFrom/TryInto<u128>`) into a `GeneralIndex` junction
+/// 2. Transfer asset back from other para-chain to local
+/// 	- with `convert_ref` to convert multilocation struct `(1,X2(ParaChain(para_id),GeneralIndex(general_index))` to a numeric asset ID
+pub struct AsPrefixedGeneralIndexFromLocalOrRemote<
+	LocalPrefix,
+	RemotePrefix,
+	AssetId,
+	ConvertAssetId,
+>(PhantomData<(LocalPrefix, RemotePrefix, AssetId, ConvertAssetId)>);
+impl<
+		LocalPrefix: Get<MultiLocation>,
+		RemotePrefix: Get<MultiLocation>,
+		AssetId: Clone,
+		ConvertAssetId: Convert<u128, AssetId>,
+	> Convert<MultiLocation, AssetId>
+	for AsPrefixedGeneralIndexFromLocalOrRemote<LocalPrefix, RemotePrefix, AssetId, ConvertAssetId>
+{
+	fn convert_ref(id: impl Borrow<MultiLocation>) -> result::Result<AssetId, ()> {
+		let local_prefix = LocalPrefix::get();
+		let remote_prefix = RemotePrefix::get();
+		let id = id.borrow();
+		let to_check_prefix = match id.parent_count() == local_prefix.parent_count() {
+			true => local_prefix,
+			false => remote_prefix,
+		};
+
+		if to_check_prefix
+			.interior()
+			.iter()
+			.enumerate()
+			.any(|(index, junction)| id.interior().at(index) != Some(junction))
+		{
+			return Err(())
+		}
+
+		match id.interior().at(LocalPrefix::get().interior().len()) {
+			Some(Junction::GeneralIndex(id)) => ConvertAssetId::convert_ref(id),
+			_ => match id.interior().at(RemotePrefix::get().interior().len()) {
```

Comment:
I would strongly suggest that you use `match_and_split` instead. I have a sense that this is implementation is error prone in edge cases.
e.g. in this case:
```Rust
let id = (1, GeneralIndex(42)).into();
let local_prefix = Here.into();
let remote_prefix = (2, Here).into();
```
I would expect your implementation to to accept the `id` while it should fail (because it matches neither of the parent counts).

Reply:
@apopiak you are absolutely right, and since `match_and_split` already do the check interior logic internally I think we can remove some check logic with more clean codes.